### ISSUE TITLE
[Request]: Add "per match confirmation" option as a special "flag"

### ISSUE THREAD

**Issue by miversen33** [MagicDuck/grug-far.nvim#24]
I have just tried out grug and its fantastic! One thing that I think would be helpful is a special flag (such as `--confirm`) that will have grug do a "Yes/No" confirmation for each file that is being replaced.

Yes I know you have the rg filters and they are fantastic (as is rg in general). But this feels like a nice QOC option.

If you are open to it, I would be happy to create a PR to implement this.

Nice work on this plugin btw!

**Comment by MagicDuck** [comment]
Thank you! Am happy people like it :)

Individual confirm sounds like a good idea. The only issue I see with it being a flag is that it is not easily discoverable. Another option, a bit more in keeping with the current paradigm to implement it as an “action” rather than a flag. Actions are things like replace, sync and close that you see help for at the top of the buffer. So essentially we would be adding a replace with confirm action. Would also be nice to scroll the buffer to the result being replaced (or move cursor there), and maybe even show some fancy extmark icon or highlight on that line to bring attention to it as user moves through the confirmations. Should also have a Cancel or Stop option. What do you think?

And yes, I am totally fine with you adding this feature and look forward to being an asshole on the code review 👅 

Let me know if you need any pointers where stuff is in the code. Currently all actions are implemented under the actions/ dir.

**Comment by miversen33** [comment]
I started poking a bit and it looks like you use ripgrep to its fullest which is great! But also means I didn't anticipate that replace was being done with ripgrep's `-r` flag (I didn't even know that flag existed!).

I will have to give implementation a think :)


> Individual confirm sounds like a good idea. The only issue I see with it being a flag is that it is not easily discoverable. Another option, a bit more in keeping with the current paradigm to implement it as an “action” rather than a flag. Actions are things like replace, sync and close that you see help for at the top of the buffer. So essentially we would be adding a replace with confirm action. 

Ya this makes sense, though I do worry that adding _too many_ actions will end up cluttering your winbar (that is a winbar right?). As an example, in split, I can't see the "Close" action. 

![image](https://github.com/MagicDuck/grug-far.nvim/assets/2640668/7f23517f-7a1f-42b0-a24b-5355a08a5c77)

I imagine adding more things to this will not improve discoverability either unfortunately. Maybe a different paradigm could be used? Something like "Mason" which has a `g?` key chord to toggle help?

![image](https://github.com/MagicDuck/grug-far.nvim/assets/2640668/20e58907-6d43-4d8a-a7e2-6431bf13b5f8)

> Would also be nice to scroll the buffer to the result being replaced (or move cursor there), and maybe even show some fancy extmark icon or highlight on that line to bring attention to it as user moves through the confirmations.

I completely agree with this :)


> Should also have a Cancel or Stop option. What do you think?

Should it? Its just a buffer and realistically confirmation will be executing a targeted `ripgrep -r`. Cancel/Stop would literally just be closing the buffer IMO

---

Thanks for the feedback and I look forward to working through this :) Lemme know your thoughts on the above

**Comment by MagicDuck** [comment]
> Ya this makes sense, though I do worry that adding too many actions will end up cluttering your winbar (that is a winbar right?). As an example, in split, I can't see the "Close" action.

Not a real winbar just virtual text :). I agree with you that it's getting large and at some point we would have to re-think and am totally fine with either a `g?` solution or a select type solution where you pick the action from a list. Actually, if I remember correctly, in nvim you can configure your picker for `vim.ui.select(...)` to be anything like telescope etc. hmmm
So we could just use that. That can be part of a separate PR though.

But yeah, I think that's a separate issue we would run into in any case at some point as more "actions" are added and should not limit us.

> Should it? Its just a buffer and realistically confirmation will be executing a targeted ripgrep -r. Cancel/Stop would literally just be closing the buffer IMO

Oh I meant, like if you are in the middle of doing a confirm replace, let's say you do the first 5 out of 10 and just want to stop the process there, so you don't get moved to the next replace match and asked to confirm again. I guess it all depends on how you implement it.
That option is available in vim a command like:
```
:%s/someting/foobar/c
```
Note the `q` option:
<img width="407" alt="image" src="https://github.com/MagicDuck/grug-far.nvim/assets/95201/65f19f76-3652-4ae0-bca5-23b52854dd94">
 
Cheers! This should be a fun feature, haha, very interactive!

**Comment by MagicDuck** [comment]
Just an idea that popped into my head, and might be useful to you (not that I am in no way saying you should implement it this way, just could be an option).

One way to get the version of the file to write on each confirm could be to:
1. add a param to fetchResults (https://github.com/MagicDuck/grug-far.nvim/blob/main/lua/grug-far/rg/fetchResults.lua) to optionally remove the `--replacement ...` part
2. as you move through the extmarks for each changed line (see similar example here: https://github.com/MagicDuck/grug-far.nvim/blob/c7c4285c28c280da5ceb0cb64eb6731f7a30cb18/lua/grug-far/actions/syncLocations.lua#L115 ),
3. if it's a file path extmark (based on the highlight group stored in the associated data in resultLocationByExtmarkId), issue a fetchResults without replacement call and a fetchResults with replacement. This will return you the sets of lines and highlight tokens for a version of the file without replacement, and the same info for the file with replacement done. The tokens have position data so you can use that to diff and obtain a version of of the file to write for each confirm.

And at this point I realize that this might break down for multiline replacements. I'll leave this here in case it ends up handy, and see my way out, haha

**Comment by serhez** [comment]
Could we have a similar system to `spectre`, where we can simply delete a line from the list of results (i.e., using `dd`) so that result won't be replaced? This will be way more convenient in many cases. For example, if you have a lot of results and just want to not replace a few, with the "confirm system" you would need to do a lot of confirming. The opposite example would be to just replace a few particular instances of a very large list of results; in `spectre` you can just replace these individually in the list of results and it's done, with the "confirm system" you would need to again do a lot of confirming. I find myself experiencing these two cases very often.

**Comment by miversen33** [comment]
I agree with this. Especially in the spirit of "treat it like a buffer"

**Comment by MagicDuck** [comment]
Yep, that sounds like a great idea! Just delete the ones you don't want and do a replace that takes into account what's in the results list!

**Comment by serhez** [comment]
I guess it would then split into two features:
1. To consider deleted lines when running the replacement command.
2. To be able to run the replacement command on a single line.

By the way, amazing work with this plugin, I've been trying it today and it's super neat, definitely a replacement for spectre :)

**Comment by MagicDuck** [comment]
Thanks! 😄 
this idea might actually make implementation a lot easier, since this is just the same functionality like `sync edits`. The only difference in sync edits is that we compare the current line in the results buffer with what was dumped in there initially so that we know which lines changed.

So in light of that we could rename `sync edits` action to just `sync`, and in the diff logic, always consider lines changed if a replacement is being done. Then we could add an additional action, `sync line` that does it only for the current line under the cursor. 
@miversen33 what do you think?

**Comment by MagicDuck** [comment]
as part of that, or in continuation, we could also change syncLocations to use libuv file read/write for async IO and better perf, see https://github.com/MagicDuck/grug-far.nvim/blob/c7c4285c28c280da5ceb0cb64eb6731f7a30cb18/lua/grug-far/actions/syncLocations.lua#L5

**Comment by MagicDuck** [comment]
the only other thing to be careful of I guess is if a replace is adding extra lines to the file. But that's probably something already missing from the current sync edits

**Comment by MagicDuck** [comment]
FYI, I've put in a change disabling sync action when --multiline flag is present, as it can lead to weird behaviour in that context.
https://github.com/MagicDuck/grug-far.nvim/pull/29

**Comment by MagicDuck** [comment]
also made files read/write use libuv async IO in https://github.com/MagicDuck/grug-far.nvim/pull/31

**Comment by MagicDuck** [comment]
since @miversen33 hasn't replied back and should be a should be a relatively simple change to unblock this type of  workflow as described in https://github.com/MagicDuck/grug-far.nvim/issues/24#issuecomment-2106330917 , I'll go ahead and take a crack at it

**Comment by miversen33** [comment]
Feel free, I was planning on poking this more this weekend. Life is busy during the week :)

**Comment by MagicDuck** [comment]
@serhez @miversen33 FYI, functionality was added in https://github.com/MagicDuck/grug-far.nvim/pull/33
It's also documented in the README. Let me know what you think...

**Comment by MagicDuck** [comment]
closing as we have this functionality now and there's been no reply. I am happy to re-open if more stuff comes up.

**Comment by serhez** [comment]
@MagicDuck thank you for the changes! Apologies for the late reply, it's been a hectic week. I tried out the new features and they are working well for me :)

**Comment by serhez** [comment]
One extra question is, why would the "Replace" command be necessary now? If I don't edit the results list, then "Replace" and "Sync All" seem to do the same thing. If I edit the results list, then running replace is likely not what I want, as it ignores my changes. Possibly I'm missing some other use case though.

**Comment by MagicDuck** [comment]
@serhez 
No worries, just being a bit OCD about closing issues, haha. Happy everything is working ok! 😄 

As you have observed, `Replace` is equivalent to `Sync All` normally, except in one situation that I know of: if you are doing multiline replacement (with `--multiline` or `--multline-dot-all` flags), as sync does not work in that case.
